File: day-2/Labfiles/03-create-plugins/Python/flight_booking_plugin.py
import os
import json
from typing import List
from dataclasses import dataclass, asdict
from semantic_kernel.functions.kernel_function_decorator import kernel_function

@dataclass
class FlightModel:
    Id: int
    Airline: str
    Destination: str
    DepartureDate: str
    Price: float
    IsBooked: bool = False


class FlightBookingPlugin:
    FILE_PATH = "flights.json"

    def __init__(self):
        self.flights: List[FlightModel] = self.load_flights_from_file()


    # Create a plugin function with kernel function attributes


    # Create a kernel function to book flights



    def save_flights_to_file(self):
        with open(self.FILE_PATH, 'w', encoding='utf-8') as f:
            json.dump([asdict(flight) for flight in self.flights], f, indent=4)

    def load_flights_from_file(self) -> List[FlightModel]:
        if os.path.exists(self.FILE_PATH):
            with open(self.FILE_PATH, 'r', encoding='utf-8') as f:
                data = json.load(f)
                return [FlightModel(**item) for item in data]

        raise FileNotFoundError(f"The file '{self.FILE_PATH}' was not found. Please provide a valid flights.json file.")

File: part-2/Labfiles/03-create-plugins/C-sharp/FlightBookingPlugin.cs
using System.ComponentModel;
using System.Text.Json;
using Microsoft.SemanticKernel;

public class FlightBookingPlugin
{
    private const string FilePath = "flights.json";
    private List<FlightModel> flights;

    public FlightBookingPlugin()
    {
        // Load flights from the file
        flights = LoadFlightsFromFile();
    }

    // Create a plugin function with kernel function attributes



    // Create a kernel function to book flights



    private void SaveFlightsToFile()
    {
        var json = JsonSerializer.Serialize(flights, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(FilePath, json);
    }

    private List<FlightModel> LoadFlightsFromFile()
    {
        if (File.Exists(FilePath))
        {
            var json = File.ReadAllText(FilePath);
            return JsonSerializer.Deserialize<List<FlightModel>>(json)!;
        }

        throw new FileNotFoundException($"The file '{FilePath}' was not found. Please provide a valid flights.json file.");
    }
}

// Flight model
public class FlightModel
{
    public int Id { get; set; }
    public required string Airline { get; set; }
    public required string Destination { get; set; }
    public required string DepartureDate { get; set; }
    public decimal Price { get; set; }
    public bool IsBooked { get; set; } = false; // Added to track booking status
}
